File: src/main/java/personal/alexgarland/helloworldmvc/controller/ErrorHandlingController.java
package personal.alexgarland.helloworldmvc.controller;

import javax.servlet.http.HttpServletRequest;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.servlet.ModelAndView;

@Controller
public class ErrorHandlingController {

  @RequestMapping("ohNoSomethingWentWrong")
  public ModelAndView renderErrorPage(HttpServletRequest httpRequest) {

    int httpErrorCode = (Integer) httpRequest.getAttribute("javax.servlet.error.status_code");

    switch (httpErrorCode) {
      case 500: {
        return new ModelAndView("serverError");
      }
      case 404: {
        return new ModelAndView("pageNotFound");
      }
    }

    return new ModelAndView("genericError");

  }

}

File: src/main/kotlin/personal/alexgarland/helloworldmvc/controller/ErrorHandlingController.kt
package personal.alexgarland.helloworldmvc.controller

import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.servlet.ModelAndView
import javax.servlet.http.HttpServletRequest

@Controller
class ErrorHandlingController {

    @RequestMapping("ohNoSomethingWentWrong")
    fun renderErrorPage(httpRequest: HttpServletRequest): ModelAndView {

        val httpErrorCode = httpRequest.getAttribute(STATUS_ATTRIBUTE) as Int

        val viewName = when (httpErrorCode) {
            500 -> "serverError"
            404 -> "pageNotFound"
            else -> "genericError"
        }

        return ModelAndView(viewName)
    }

    companion object {

        private const val STATUS_ATTRIBUTE: String = "javax.servlet.error.status_code"

    }

}

